MAINT: struct assignment "by field position", multi-field indices return views

ahaldane opened the pull request:
This is a try at improving structured array assignment, aiming for 1.11. The goal is to remove various inconsistencies in structure assignment, and to allow multi-field indexing to return a view.

The biggest change is that assignment from one structured array to another now works "by position" instead of "by field name", as in this example:

```
>>> arr1 = np.array([(1,2,3)], dtype=[('a', 'i4'), ('b', 'i4'), ('c', 'i4')])
>>> arr2 = np.array([(4,5,6)], dtype=[('b', 'i4'), ('a', 'i4'), ('c', 'i4')])
>>> arr1[:] = arr2
```

Previously, arr1 would be set to "(5,4,6)" but with this PR it is set to "(4,5,6)". 

This fixes the issues in structured array assignment outlined below, and allows casting/assignment in new situations. Then I allow multi-field indexing (like `arr1[['b', 'c']]`) to return a view instead of a copy (issue #5994), since assignment to this view will now work more sensibly.

This PR also fixes #2353 #6085 #3351.
### Problems in structure assignment

There are two code paths used to assign structured items, 1. using `setitem` or 2. using the `dtype_transfer` functions. The two paths behave differently from each other, and each sometimes have unexpected behaviors. Consider this setup:

```
def testassign(arr, v):
    c = arr.copy()
    c[0] = v  # assign using setitem
    c[1:] = v # assign using "dtype_transfer"
    print(c)

arr = ones(2, dtype=[('foo', 'i8'), ('bar', 'i8')])
v1 = array([(2,3)], dtype=[('foo', 'i8'), ('bar', 'i8')])
v2 = array([(2,3)], dtype=[('bar', 'i8'), ('foo', 'i8')])
v3 = array([(2,3)], dtype=[('bar', 'i8'), ('baz', 'i8')])
v4 = array([(2,)],  dtype=[('bar', 'i8')])
v5 = array([(2,3)], dtype=[('foo', 'f8'), ('bar', 'f8')])
w = arr.view({'names': ['bar'], 'formats': ['i8'], 'offsets': [8]})

testassign(arr, v1)
testassign(arr, v2)
testassign(arr, v3)
testassign(arr, v4)
testassign(arr, v5)
w[:] = 4
print arr
```

Output:

```
[(2, 3) (2, 3)]
[(2, 3) (3, 2)]
[(2, 3) (0, 2)]
[(2, 0) (0, 2)]
[(4607182418800017408, 4611686018427387904) (2, 3)]
[(140514438937232, 4) (140514438937232, 4)]
```

In summary, only the assignment using exactly the same dtype (v1) works, all other cases are broken in some way. For v2, `setitem` effectively assigns the fields "by position", but `dtype_transfer` code assigns "by field name". For v3, `dtype_transfer` converts "missing" fields to 0. In v4 `dtype_transfer` sets missing fields to 0, while `setitem` pads short buffers with 0. For v5, `setitem` copies the bytes directly (no casting). For the "partial view" `w`, we get uninitialized values at unviewed fields in both paths.
### Proposed Solution

In this PR I made all field assignment occur "by position", and cast if needed. That is, if you order the fields from left to right in dst and src arrays, dst field 0 gets set to src field 0 and so on, regardless of field name. This makes sense since the "python representation" of a structured element (in `VOID_getitem`) is a tuple, and tuple assignment to a structure works "by position" too:

```
>>> a[0] = (1,2)
>>> a[0] = a[1].item() #works the same, essentially "a.setitem(0, a.getitem(1))"
```

Assignment to a different number of fields (v4 above) is no longer allowed. With this behavior, the output of the example will be:

```
[(2, 3) (2, 3)]
[(2, 3) (2, 3)]
[(2, 3) (2, 3)]
# v4 raises exception
[(2, 3) (2, 3)]
[(1, 4) (1, 4)]
```

This is a backward incompatible change, but I suspect it's not a big problem for most people given how inconsistently structure assignment worked. The most common case - assigning using an identical dtype (v1) - behaves the same as before. I ran unit tests for scipy, pandas and astropy with the PR on my system (python2, x64 linux). Scipy and pandas are unaffected, and astropy fails 1 new test ([this one](https://github.com/astropy/astropy/blob/master/astropy/units/tests/test_quantity_array_methods.py#L557)). 

One alternative is to make all assignment work "by field name", but that also involves a compatibility break for all of the setitem cases. The fact that `dtype_transfer` code assigns by field name is currently undocumented (as far as I see), however there are a couple unit tests that test whether "reordering" of fields is allowed (like in v2), which depends on assignment by field name. I removed/modified these tests. I think assignment by field name is confusing and bug prone though, and without it I was able to get rid of a number of unit tests for edge cases we no longer have to worry about (eg, checking that "unviewed" fields of various kinds got set to 0 or None).

Historical note: Based on some comments in unit tests it looks like numpy 1.5 and before assigned fields "by position", but this was changed in 1.6 to assignment "by field name" in PR #36, though this was not noted anywhere in docs or release notes.

Implementation involved rewriting parts of `VOID_setitem` and `get_fields_transfer_function`. 
### Multi-Field Assignment

With these changes it is now reasonable to allow writeable multi-field views, which allows assignment like:

```
>>> a = np.ones(4, dtype=[('foo', 'i8'), ('bar', 'i8'), ('baz', 'f4')])
>>> a[['bar', 'baz']] = (2, 3.0)
>>> a[['baz', 'foo']][:2] = (4.0, 5)
>>> a
array([(5, 2, 4.0), (5, 2, 4.0), (1, 2, 3.0), (1, 2, 3.0)], 
      dtype=[('foo', '<i8'), ('bar', '<i8'), ('baz', '<f4')])
```

Note that with assignment "by position" reordering the fields (like in v2) would no longer work. However, multi-field assignment makes up for this, since fields will be assigned in the order they are subscripted, allowing reordering in a different way. For example, 

```
>>> a[['foo', 'bar']] = a[['bar', 'foo']]
```

swaps the field data (while it was a no-op/broken in the old assignment "by field name").

Note that multi-field views (whether by this PR or another) introduce another backward incompatible change, since the byte-arrangement of a multi-field copy is different from the byte-arrangement of a multi-field view:

```
>>> a = ones(10, dtype='i8,i8,i8')
>>> a[['f0','f2']].dtype
# old ouput, for the multi-field copy:
dtype([('f0', '<i8'), ('f2', '<i8')])
# this PR's output, for multi-field view
dtype({'names':['f0','f2'], 'formats':['<i8','<i8'], 'offsets':[0,16], 'itemsize':24})
```

This is a problem for code which does things like `a[['f0','f2']].view('i4')` (like in unit test `test_field_names`), which would end up viewing different values in the old vs new code. In the new code this particular view raises an error as a result of my other PR #4488 to add safety checks to views. Similarly, `np.recarray` code uses such a view internally, and needed to be fixed. There might be a similar problem for the planned conversion to `np.diagonal` to a view.

I suggest we add a deprecation warning for code that attempts such a view in 1.10. Multi-field views have been planned for a long time and there is already a deprecation warning that copies will become views, but the warning only appears if you try to assign to the copy, and not if you try to view the copy. I've made another PR, #6054, which adds the appropriate deprecation warnings.

This change required a small edit to `_internals.py`, and removing warnings in `mapping.c`.
### Other changes

1) One difference between `setitem` and `dtype_transfer` was that `dtype_transfer` allows assignment of a scalar to a structure (`a[ind] = 1`), which sets all fields to that value, while `setitem` did not. This is important for eg `np.ones`, which initializes the array with such a statement. I made it so both paths now allow assignment using a scalar.

2) Relatedly, got rid of undocumented special case for assignment from a structure to a scalar array of boolean type: If the scalar was of boolean type it would be set to "True" regardless of the values in the structured array, with the explanation that "the existence of fields implies True". This broke a MaskedArray test, which I fixed by updating `make_mask` to do the equivalent.

3) Previously, if you tried to assign a structured array to a non-structured array just the first field was copied over. Eg

```
 >>> a = np.array([1,2,3])
 >>> b = np.array([(10,20,3.0)], dtype='i4,i8,f4')
 >>> a[:] = b
 >>> a
 array([10, 10, 10])
```

Now, I made it raise an exception unless the structured type has only one field, since I found it confusing.

4) Slightly changed PyArray_EquivTypes for structured arrays. Now it checks that the fields are in the same order in both dtypes. (previously only cared that the dtype.fields dicts had the same values). In other words, field order is now important!

5) I rewrote most of the docs on structured arrays. A possibly iffy part of the new docs is the section "obsolete features", in which I discourage use of field titles and use the "second" dictionary based form of dtype specification.

ahaldane committed:
MAINT: improve struct assignment, allow multi fields views

This commit attempts to make structure assignment more consistent, and
then changes multi-field indices to return a view instead of a copy.

Assignment between structures now works "by field position" rather than
"by field name".

Fixes #2353, fixes #6085, fixes #3351, fixes #6085, fixes #6314,
fixes #2346, fixes #7058, fixes #3641, fixes #5994, fixes #7262,
fixes #7493

ahaldane committed:
TST: Fix/Remove old unit tests for structure assignment

ahaldane committed:
TST: Add new unit tests for structure assignment

ahaldane commented:
All right, this PR is complete except for release notes.

Although I'm aiming for 1.11 which is a long way off, it might be worth discussing the deprecation warnings I mentioned into 1.10. 

Also, for some context [here are a few notes](https://gist.github.com/ahaldane/2bfd2293affcb9cff68b) on other things that could be improved about structured arrays.

homu commented:
:umbrella: The latest upstream changes (presumably #6269) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
@ahaldane Needs a rebase.

charris commented on doc/sphinxext:
```diff
@@ -1 +1 @@
-Subproject commit ef988a4a4658c991f4445f6241ab02d74710c6e3
+Subproject commit 84cc897d266e0afc28fc5296edf01afb08005472
```
This probably needs `git submodule update.`

charris replied:
The replaced hash is the correct one.

ahaldane replied:
Yeah I was just trying to figure out what is going on here. Not sure why it got changed, but I think I fixed it.

ahaldane commented:
Rebased. Current summary:

 1. In `dtype_transfer.c`: Modified the dtype-transfer code so that assignment between structured arrays works "by field position". Also removes 2 special cases involving assignment to scalar arrays (see previous comment). 
 2. In `arraytypes.c.src`: Modifed the element assignment code so assignment to structured arrays works "by field position". Likewise in `scalartypes.c.src` allow multi-field subscripts of structured scalars.
 3. In `mapping.c`: Made multi-field subscripts of an array return a view rather than a copy.
 4. Smaller fixups: 
   * In `array_assign_array.c`, make sure copying a view of structured array to itself (like `arr[['a', 'b']] = arr[['b', 'a']]`) uses a temp buffer (didn't matter before since couldn't make such views).
   * In `multiarraymodule.c`, make sure `_equivalent_fields` checks that the fields of the two datatypes have the same order too: Field order now matters!
   * In MaskedArray module: Small tweak to account for removal of the boolean special case in `dtype_transfer.c`.

Chopped out the DOC commits for now, and put the proposed new docs [on gist](https://gist.github.com/ahaldane/7d1873d33d4d0f80ba7a54ccf1052eee) (to replace [the old ones](http://docs.scipy.org/doc/numpy-1.10.1/user/basics.rec.html)). I can submit them as a separate PR later.

Fixes #2353 #6085 #3351 #6085 #6314 #2346 #7058 #3641 #5994 #7262 #7493 with the following caveats:

#2353: Assignment now casts properly but that particular assignment is now disallowed: The user must do `arr[0] = tuple(np.array([10,20]))` instead of `arr[0] = np.array([10,20])`. I suppose I could relax this if desired.

#3351: Instead of `U[0] = np.eye(4)`, the user must do either `U[0] = (np.eye(4),)` or `U[0]['x'] = np.eye(4)` or `U['x'][0] = np.eye(4)`

#7058: Assignment is not allowed between a subarray and a structured element, so the `astype` example raises an error. However `view` works, which is what the user wanted anyway.

homu commented:
:umbrella: The latest upstream changes (presumably #6430) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
@ahaldane probably need to look at getting this into 1.12. Could you take care of the rebase and take a rested look at the code?

charris commented:
```
>>> a[['foo', 'bar']] = a[['bar', 'foo']]
```
Does this actually work if the RHS is a view? Are all multifield "views" actually views if they don't refer to contiguous positions in the dtype?

ahaldane commented:
@charris I'll take care of the rebase sometime this week.

And yes, it works and both sides are views. They are real views, just as much as a structured array which contains padding bytes between its fields.

charris commented:
Guess I'm confused. I would think elements would get overwritten before reference, or is it just views that are exchanged?

ahaldane commented:
I think I see your concern. In `array_assign_array.c` (at the top of the diff) I check if `arrays_overlap` and use a temporary buffer if so. 

So the self-swapping case involves using a temp buffer to avoid overwriting before reference. (slow but it works).

charris commented:
There is more to this patch than I thought ;)

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
-        int savedflags;
+    flags = PyArray_FLAGS(ap);
+    if (PyDataType_HASFIELDS(descr)) {
+        /* copy field-by-field in the cases src is:
```
First line blank.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
-            return -1;
+        npy_int i, names_size;
+        npy_intp offset;
+
```
I'd probably drop these blank lines in the declarations.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
+         * to create a buffer array which we then clobber.
+         */
+        Py_INCREF(descr);
+        buffer = (PyArrayObject*)PyArray_NewFromDescr(&PyArray_Type,
```
Need an error check.

charris replied:
Note that if `descr` is subarray, then it will be decrefed even on success, so you can not rely on using it later on, which you do.

charris replied:
See #7757. If you plan to reuse `descr` after this call you need another incref and then decref before function exit, but after use.

charris replied:
Mind, it might be OK if something else is already holding a reference, just needs to be checked.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
+            }
+
+        }
+        else{
```
Space after `else`.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
+        if (PyTuple_Check(op)) {
+            /* if it's a tuple, copy field-by-field to buffer, then to dst */
+
+            if(names_size != PyTuple_Size(op)){
```
Space after `if` and before `{`.

charris commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2531,22 +2532,79 @@ get_fields_transfer_function(int aligned,
+        src_names = src_dtype->names;
+        if (PyTuple_GET_SIZE(src_names) != 1) {
+            PyErr_SetString(PyExc_ValueError, "Can't cast from structure to "
+             "non-structure, except if the structure only has a single field.");
```
Indentation. It probably break the preceding line after the first `,`.

charris commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2531,22 +2532,79 @@ get_fields_transfer_function(int aligned,
+        key = PyTuple_GET_ITEM(src_names, 0);
+        tup = PyDict_GetItem(src_dtype->fields, key);
+        if (!PyArg_ParseTuple(tup, "Oi|O", &src_fld_dtype,
+                                                &src_offset, &title)) {
```
Indentation.

charris commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2531,22 +2532,79 @@ get_fields_transfer_function(int aligned,
+        }
+
+        if (PyArray_GetDTypeTransferFunction(0,
+                                src_stride, dst_stride,
```
Indentation. Align arguments vertically on `0` here.

charris commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2611,288 +2669,74 @@ get_fields_transfer_function(int aligned,
-                                &fields[field_count].stransfer,
-                                &fields[field_count].data,
+        if (PyArray_GetDTypeTransferFunction(0,
+                                src_stride, dst_stride,
```
See above on alignment of arguments.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
+                return -1;
+
+            }
+            if (PyArray_TYPE(oparr) == NPY_OBJECT){
```
Space before `{`.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
+            int ret;
+
+            /* Fast path if dtypes are equal */
+            if(PyArray_EquivTypes(srcdescr, descr)){
```
Space after `if`, before `{`.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
+            /* Slow path */
+            ret = PyArray_CastRawArrays(1, srcdata, ip, 0, 0,
+                                        srcdescr, descr, 0);
+            if (ret != NPY_SUCCEED){
```
Space before `{`.

charris commented:
Can't say I've verified the functionality, but there are various whitespace style errors that need fixing. The oddities of `PyArray_NewFromDescr`, i.e., don't use the `descr` argument after unless you hold a reference needs to be looked at. I suspect this isn't the only place in numpy that this problem lurks, but it only appears when `descr->subarray` is true.

Anyway, make a comment when ready so that github will ping.

ahaldane commented:
Updated.

I decided to get rid of the buffering in `VOID_setitem` entirely, so the issues with `PyArray_NewFromDescr` are gone. The original motivation for the buffer was to avoid situations where half of the fields of an element got filled in and then failure occurred, but actually I don't think allowing  that is too much of a problem because there are similar cases where numpy feels free to clobber arrays after an error. Eg,

    >>> a = np.ones(2, dtype='S3')
    >>> class F:
    ...    def __str__(self):
    ...        raise Exception()
    ...
    >>> a[:] = ['asd', F()]
    Exception:
    >>> a
    array(['asd', '1'], 
          dtype='|S3')

charris commented:
I suppose a warning that the array may be clobbered could be part of the exception message. Let me know when you think this is ready.

ahaldane commented:
I think it's ready, thank you for reviewing!

I didn't set a special error message about clobbering, since there will already be an exception set if there was a problem.

Also, we may not want this for 1.12 - it might need a round of deprecation warnings as implemented in #6054, and then this PR can go in 1.13.

charris commented:
Hmm, I was thinking that 1.12 would be difficult anyway, so the trade off is getting #6054 tested vs deprecation. If you think current usage downstream requires a deprecation, or that that would be the smoother way forward, I'll go with that. Guess I was thinking of this as more of a bug fix than an enhancement/change.

ahaldane commented:
I'd be more comfortable with a deprecation cycle. 

The main dangerous change is that taking a view of a multi-field index will behave quite differently: Multi-field indexing returned a copy before, with had a different memory layout from the view this PR returns.

charris commented:
@ahaldane  I've moved this on to 1.13.0 and will concentrate on the deprecation PR. We may want to move it on further following our two releases policy.

ahaldane commented:
@charris sounds good to me

homu commented:
:umbrella: The latest upstream changes (presumably #6054) made this pull request unmergeable. Please resolve the merge conflicts.

mhvk commented:
@ahaldane - there's no hurry, but you mentioned at the very top that one astropy `Quantity` test failed. The test file has changed in the meantime, so I'm not 100% sure which test it was, but given that there really is only one with `recarray`, it seems almost certain it is that one ( https://github.com/astropy/astropy/blob/master/astropy/units/tests/test_quantity_array_methods.py#L543). The odd thing is that I cannot see how that test would be broken by your PR.  Anyway, I may test it myself later, but thought I would ask in case you had any insight.

ahaldane commented:
@mhvk I don't remember which test it was, but after I rebase (sometime soon I hope...) I'll test it again. I vaguely remember it seemed easily fixable.

charris commented:
@ahaldane Do you want to get this into 1.13?

ahaldane commented:
Yes, very much!

ahaldane commented:
Rebased for 1.13.  Tests pass, I think, not sure why it says I cancelled the appveyor test.

charris commented:
I canceled the test. Appveyor had a problem Sunday, and after the fix was rerunning over a hundred tests, so I canceled them rather than waiting for several days for it to catch up. I'll restart this one now that it is mergeable.

homu commented:
:umbrella: The latest upstream changes (presumably #8590) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -1613,7 +1613,11 @@ def make_mask(m, copy=False, shrink=True, dtype=MaskType):
-            result = np.array(m, dtype=dtype, copy=copy)
+            # existence of fields implies True
+            if m.dtype.fields and np.dtype(dtype) == np.bool_:
+                result = np.ones(m.shape, dtype=dtype)
```
This doesn't seem right to me. I'm assuming the aim here is to convert a per-element mask to a per-field mask? This is converting a per-element mask to a "everything is masked" mask

ahaldane replied:
This was necessary in order to preserve the old behavior and get the MA tests to pass, at the time I wrote it. It is possible that with the recent changes to structured masks you've worked on this can be updated.

More specifically, in this PR I removed a special case in the C code [here](https://github.com/numpy/numpy/pull/6053/files#diff-d3108a3599564ada8ce08fe45f539350L2650) since I totally rewrote the casting code for structs. 

That block is a special case that converts strucs to booleans, as in:
```python
>>> a = np.zeros(4, dtype='i4,f4')
>>> a.astype(bool)
array([ True,  True,  True,  True], dtype=bool)
```
I removed it because it didn't fit in with my changes and seemed like a weird hack.

However that behavior was necessary in MA code somewhere, so I had to patch up the MA code as in the the line you commented on.

ahaldane replied:
I just tried rebasing, removing this block in favor of the code from #8590, but I get a test failure with trace:
```
Traceback (most recent call last):
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/ma/tests/test_core.py", line 4041, in test_make_mask
    test = make_mask(mask)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/ma/core.py", line 1610, in make_mask
    result = np.array(filled(m, True), copy=copy, dtype=dtype, subok=True)
ValueError: Can't cast from structure to non-structure, except if the structure only has a single field.
```

I don't have time to debug right now, but I'll get around to it sometime this week.

ahaldane replied:
I couldn't help myself and ended up doing some debugging...

I think there is still a problem with MA masks after #8590, specifically with the `make_mask` method. This seems wrong:
```python
>>> a = np.zeros(4, dtype='i4,f4')
>>> np.ma.make_mask(a)
array([ True,  True,  True,  True], dtype=bool)
```
Shouldn't it return 
```python
array([(False, False), (False, False), (False, False), (False, False)], 
      dtype=[('f0', '?'), ('f1', '?')])
```
?

Actually, `make_mask` did that before #8590 too.

eric-wieser replied:
It does what you'd expect if you pass `np.ma.make_mask(a, dtype='b,b')` on 1.11, but fails even then on master.

charris commented:
Needs another rebase.  @eric-wieser Comments?

charris commented:
I'm going to push this off to 1.14.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,193 @@ VOID_getitem(void *input, void *vap)
+ * numpy structure, in VOID_setitem.
+ */
+NPY_NO_EXPORT int
+_setupField(int i, PyArray_Descr *descr, PyArrayObject *arr,
```
I think `_snake_case` is the convention we use elsewhere for helper functions?

ahaldane replied:
true, done.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -755,66 +756,193 @@ VOID_getitem(void *input, void *vap)
-            if (res < 0) {
-                break;
+            return 0;
+        }
```
This `if` statement might read better as a helper function as then it can appear in the `PyArray_Check(op)` and `PyArray_IsScalar`  branches above, and the whole function becomes a single chained if/else

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -843,29 +971,24 @@ VOID_setitem(PyObject *op, void *input, void *vap)
-            PyErr_SetString(PyExc_ValueError,
-                    "Setting void-array with object members using buffer.");
-            return -1;
-        }
```
Why is this no longer an issue?

ahaldane replied:
Because all cases involving objects now must fall into the HASOBJECTS branch. Because of other work I've done I'm fairly sure it is not possible to create a void unstructured type containing pointers.

In the old code, however, many structured arrays would fall into this part of the code (eg, see #6314).

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -13,6 +13,7 @@
 #include "npy_pycompat.h"
 #include "numpy/npy_math.h"
 #include "numpy/halffloat.h"
+#include "lowlevel_strided_loops.h"
```
At a guess from the github diff, Is this for `_unpack_field`?

ahaldane replied:
yes, but looks like it is already included a few lines below so I'm removing this line.

 (perhaps it was added in another PR)

eric-wieser commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2520,7 +2521,7 @@ _strided_to_strided_field_transfer(char *dst, npy_intp dst_stride,
 /*
  * Handles fields transfer.  To call this, at least one of the dtypes
- * must have fields
+ * must have fields. Does not take care of object<->structure conversion
```
What do `object` and `structure` mean here?

ahaldane replied:
As I recall, this refers to converting object arrays (or python objects) to structures. 

If you go to the (only) place that `get_fields_transfer_function ` is called, note that it is in an if-block that excludes NPY_OBJECT types.

ahaldane commented:
Updated. The holdup was what to do about the `numpy/ma/core.py ` code. I've now decided to add a special case in `make_mask` in order to reproduce the old behavior. 

Note the old behavior was itself a funny special case in the C code, [here](https://github.com/numpy/numpy/pull/6053/files#diff-d3108a3599564ada8ce08fe45f539350L2650) as noted before, I've just moved it to python now.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -772,66 +772,197 @@ VOID_getitem(void *input, void *vap)
+            PyArray_Descr *srcdescr = ((PyVoidScalarObject*)op)->descr;
+            char *srcdata = ((PyVoidScalarObject*)op)->obval;
+            if (!PyDataType_HASFIELDS(srcdescr)) {
+                PyErr_SetString(PyExc_TypeError, msg_nostruct);
```
Would it make sense to move this error checking into the helper funciton too?

ahaldane replied:
yes!

eric-wieser commented:
@charris:

> I'm going to push this off to 1.14.

We committed in the previously release notes (and possibly some warning messages) to bringing this to 1.13 - so at the very least, we need to fix these warnings, and add a release note explaining that the feature is delayed

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -772,66 +772,197 @@ VOID_getitem(void *input, void *vap)
+
+    /* Slow path */
+    ret = PyArray_CastRawArrays(1, srcdata, dstdata, 0, 0,
+                                srcdescr, dstdescr, 0);
```
What type of cast is even allowed here?

ahaldane replied:
This accounts for assigning between structures with different field types, eg ''f4,f4' to 'i4,i4'.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1418,26 +1418,30 @@ array_putmask(PyObject *NPY_UNUSED(module), PyObject *args, PyObject *kwds)
  * Return 1 if the contents are the same, 0 if not.
```
This comment should be updated to explain the new stricter conditions

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2204,27 +2204,18 @@ voidtype_subscript(PyVoidScalarObject *self, PyObject *ind)
-    PyErr_SetString(PyExc_IndexError, "invalid index");
-    return NULL;
+    /* if it's not a number, try ndarray indexing */
+    args = Py_BuildValue("(O)", ind);
```
Does changing this mean that `()` and `...` are now treated as valid indices?

If so, should these not be valid indices whether or not `self` has fields?

ahaldane replied:
good idea, I'll add that

eric-wieser reviewed (commented):
Struggling to review `dtype_transfer.c` due to size, but the rest looks good.

Not too happy about the clobbering going on in `_setup_fields` - is this thread-safe (and does it matter if not)? Could we instead just construct a single temporary view, and shift it across each field in turn?

ahaldane commented:
Updated. The changes are:

1.  moved an `if` statement to the `_copy_and_return_void_setitem` helper function
2.  update the comment for `equivalent_fields`
3. allowed unstructured void scalars to be indexed with `()` and `...`, plus a test for this.

 Note that indexing a 'V4' scalar with `...` gives back a possibly weird-looking buffer inside an array (as does `np.array(np.void(4))`), but I think that is outside the scope of this PR.

ahaldane commented:
Also, the clobbering going on in `_setup_fields` is a long-standing wart. Search `arraytypes.c.src` for the word `TODO` and you will see there is a lot of similar clobbering going on throught this file.

@jaimefrio already tried to minimize the problem by introducing the `dummy_struct`s where possible a year ago, so we clobber a dummy instead of the actual array, but this wasn't possible everywhere.

As I recall, the underlying problem is that all of these functions accept only an array `ap` as input, when probably they should also accept flags and a dtype as further arguments.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2191,40 +2191,33 @@ voidtype_item(PyVoidScalarObject *self, Py_ssize_t n)
-    return NULL;
+    /* if it's not a number, try ndarray indexing */
+    args = Py_BuildValue("(O)", ind);
+    ret = gentype_generic_method((PyObject *)self, args, NULL, "__getitem__");
```
Can't this just be `gen_arrtype_subscript`?

ahaldane replied:
Now that you mention it, I am going to change it to using  `__getitem__` even in in the unstructured case a few lines above too (will update). This better accounts for subclasses of `void` like `mvoid` which redefine `__getitem__`.

Using `gen_arrtype_subscript` actually breaks maskedarray printing, eg with your suggested change I get:
```python
>>> a = np.ma.array(np.zeros(2, dtype='i4,i4'), mask=[(0,0),(1,0)])
>>> a
masked_array(data = [(array(0, dtype=object), array(0, dtype=object))
 (array(--, dtype=object), array(0, dtype=object))],
             mask = [(False, False) ( True, False)],
       fill_value = (999999, 999999),
            dtype = [('f0', '<i4'), ('f1', '<i4')])
```

I will note that with the current PR (using `__getitem__` instead of  `gen_arrtype_subscript`), mvoid types don't work properly:
```python
>> a = np.ma.array(np.zeros(2, dtype='i4,i4'), mask=[(0,0),(1,0)])
>>> b = a[1]
>>> b[()]
masked
```
this is because `mvoid.__getitem__` isn't properly implemented to account for tuple indexing, since [this](https://github.com/numpy/numpy/blob/73c15f7b30658cbedc9c42e59b818b3d77dbc2c3/numpy/ma/core.py#L6034) line tests `m[indx]`, which returns a void `(1,0)` in this case. But honestly I don't want to try to fix maskedarray stuff in this PR if that's OK.

eric-wieser commented:
> Note that indexing a 'V4' scalar with ... gives back a possibly weird-looking buffer inside an array

I think that the keyword is "looking", and the bug is actually in `arr_repr` there - the result seems to other behave as I would expect

eric-wieser commented on numpy/core/tests/test_indexing.py:
```diff
@@ -106,6 +106,11 @@ def test_empty_tuple_index(self):
+    def test_void_scalar_empty_tuple(self):
+        s = np.zeros(1, dtype='V4')[0]
+        assert_equal(s[()].dtype, s.dtype)
+        assert_equal(s[()], s)
```
No test for `...`?

eric-wieser commented on numpy/core/tests/test_indexing.py:
```diff
@@ -106,6 +106,11 @@ def test_empty_tuple_index(self):
         assert_(isinstance(a[()], np.int_))
 
+    def test_void_scalar_empty_tuple(self):
+        s = np.zeros(1, dtype='V4')[0]
```
Or better yet, just `np.zeros((), dtype='V4')`

ahaldane commented:
Updated. The main change was tweaks to the `voidtype_subscript` code. In the end using neither `gen_arrtype_subscript` nor a call to ` __getitem__` (as discussed above) was suitable to get tuple and ellipsis indexing to work, instead I found a simple custom solution.

`gen_arrtype_subscript` isn't suitable because 1. it always returns a scalar and we don't want that for the ellipsis case, and 2. because it raises an `IndexError` instead of a `ValueError`, as is expected in various tests.

` __getitem__` wasn't suitable because it always returns an array, unless we wrap it in PrArray_Return in which case it always returns a scalar. Plus I decided it doesn't make sense to call subclass code here.

Instead I explicitly check for ellispsis, which I assume is the only case where an 0d array can be returned.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -772,66 +772,194 @@ VOID_getitem(void *input, void *vap)
-                PyArray_CLEARFLAGS(ap, NPY_ARRAY_ALIGNED);
+            if (PyArray_TYPE(oparr) == NPY_OBJECT) {
+                return PyArray_SETITEM(vap, input,
+                            PyArray_GETITEM(oparr, PyArray_DATA(oparr)));
```
What's the code that triggers this path? An object array where each item is a void scalar? Can we just invoke `VOID_setitem` directly here>

ahaldane replied:
It is for when the src is an object array of size 1. 

But actually I now realized I can remove this special case, as long as I also remove the special if-statement  for structured types in `_copy_and_return_void_setitem`. Will update.

eric-wieser commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2611,313 +2670,74 @@ get_fields_transfer_function(int aligned,
-        }
-        field_count = 0;
-        /*
-         * Special case bool type, the existence of fields implies True
```
I don't know if we can actually remove this without a deprecation cycle

ahaldane replied:
Perhaps. But (famous last words) I think it is such an obscure special case I don't think it will affect anyone.

(The fact that MA code depended on this behavior in one place might say otherwise, but I still feel pretty confident about that).

To clarify the discussion, this changes the fact that in numpy master
```python
>>> a = np.zeros(4, dtype='bool')
>>> b = np.zeros(4, dtype='i4,i4')
>>> a[:] = b
>>> a
b = np.zeros(4, dtype='i4,i4')
```

In this PR, this code now raises a ValueError.

eric-wieser commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2531,22 +2532,80 @@ get_fields_transfer_function(int aligned,
+     */
+
+    /* 1. dst is non-structured. Allow transfer from single-field src to dst */
+    if (!PyDataType_HASFIELDS(dst_dtype)) {
```
In the interest of reducing the diff here, can we swap the order of 1. and 2.?

ahaldane replied:
Done.

eric-wieser commented:
This changset makes the `np.dtype(dict(a=(int, 0), b=(int, 0)))` constructor dangerous, since "field position" is non-deterministic. So that constructor needs to be made to sort names alphabetically

ahaldane commented:
Correct. Check out my proposed new docs for structured arrays (here)[https://gist.github.com/ahaldane/7d1873d33d4d0f80ba7a54ccf1052eee#second-dictionary-based-dtype-specification] where I note this. You have a good idea to make is sort alphabetically though, I was just going to warn the user not to use that constructor.

I'll submit those docs as a PR to accompany this one. (I'll get back to all my PRs shortly, I just have a lot of work this week).

mhvk commented:
Does one really want to sort alphabetically? As of python 3.6, `dict` is insertion-ordered... I think "don't use it" or "use `OrderedDict`" would be more useful than enforcing an ordering which may well be wrong.

homu commented:
:umbrella: The latest upstream changes (presumably #9060) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
Needs a rebase and release note.

ahaldane commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -2638,286 +2640,104 @@ get_fields_transfer_function(int aligned,
         }
-        field_count = 0;
-        /*
-         * Special case bool type, the existence of fields implies True
```
Note discussion of removal of this section [here](https://github.com/numpy/numpy/pull/6053#discussion_r123046754).

ahaldane commented:
Rebased, added release note, and made a few small tweaks to address comments.

Also, I looked through `numpy/lib/recfunctions.py` for more problems similar to the problem found in #9070, but didn't notice any more conflicts with this PR. (I did notice other problems in that file though, specifically the frequent incorrect use of `dtype.descr`).

charris commented:
Needs rebase. Where are we with this now? Are there adjustments needed in light of the overlap detection added in 1.13 or is that lower down the stack than this?

charris commented on doc/release/1.14.0-notes.rst:
```diff
@@ -61,3 +61,40 @@ integer scalars are now unaffected by ``np.set_string_function``
+identically-named field in the source array or to 0 if the source did not have
+a field.
+
+Other changes to structure array manupilation
```
"Other" is sort of a hanging reference here. Is it possible to be more specific without a multiline heading?

ahaldane commented:
I'll get to this soon, this weekend I think.

ahaldane committed:
BUG: account for field titles in multi-field indexes

Fixes #9625

ahaldane commented:
Rebased, modified release note, an added an extra commit to take care of field titles, to fix #9625. 

This PR works with the overlap changes of 1.13 (which were partly reverted anyway).

ahaldane commented:
Also, the last commit for field-titles disables two previously-allowed but buggy behaviors:

Before:
```python
>>> a = np.zeros(4, dtype=[(('title', 'b'), 'i'), ('c', 'i')])
>>> a[['title', 'c']]  # weirdly makes title into field name
array([(0, 0), (0, 0), (0, 0), (0, 0)],
      dtype=[('title', '<i4'), ('c', '<i4')])
>>> a[['b', 'c']]  # loses title
array([(0, 0), (0, 0), (0, 0), (0, 0)],
      dtype=[('b', '<i4'), ('c', '<i4')])
>>> a[['b', 'b']]
ValueError: name already used as a name or title
```

Now: 
```python
>>> a = np.zeros(4, dtype=[(('title', 'b'), 'i'), ('c', 'i')])
>>> a[['title', 'c']]
KeyError: 'cannot use multi-index with field titles'
>>> a[['b', 'c']]  # note title is retained
array([(0, 0), (0, 0), (0, 0), (0, 0)],
      dtype=[(('title', 'b'), '<i4'), ('c', '<i4')])
>>> a[['b', 'b']]
KeyError: 'duplicate field of name b'
```

charris commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2179,35 +2179,31 @@ static PyObject *
+     */
+    ret = array_subscript((PyArrayObject *)res, ind);
+    Py_DECREF(res);
+    return PyArray_Return((PyObject*)ret);
```
`PyArray_Return` takes a pointer to `PyArrayObject`.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -774,66 +774,172 @@ VOID_getitem(void *input, void *vap)
+            if (names_size != PyTuple_Size(op)) {
+                errmsg = PyUString_FromFormat(
+                        "could not assign tuple of length %d to structure "
+                        "with %d fields.", PyTuple_Size(op), names_size);
```
Incorrect format specifier for `names_size`. You can use `NPY_INTP_FMT`.
```
"with " NPY_INTP_FMT " fields."
```

charris commented:
Couple of compiler warnings need fixes.

ahaldane commented:
Fixed the compiler errors, but there also appears to be a refcounting error in my last commit.. I'll fix it soon.

ahaldane commented:
trying to get tests to run...

charris commented:
Sometimes you just gotta close and reopen. Ill try that.

charris commented:
Seems github was being slow to deliver notifications. At least that was what ci status seems to be saying.

charris commented:
@ahaldane Still need refcounting fix?

ahaldane commented:
Nope, it's fixed. 

I was doing `Py_DECREF(name)` after calling `PyUString_ConcatAndDel(&errmsg, name);`, fixed now.

charris commented:
OK, let's get this in. Thanks Allan.

ahaldane commented:
Woohoo! Thanks Chuck and reviewers.

aragilar commented:
Is this supposed to break dtypes where the offsets are not monotonically increasing? h5py requires the ability to specify out of order offsets in dtypes.

ahaldane commented:
No, there is still no constraint on the offsets.

The only thing that has changed is that the order of the field names (as listed by `dtype.names`)  now matters in some cases, such as when testing equality, and during assignment. Previously, it did not, though I suspect many people assumed it did.

This is the new 1.14 behavior:
```python
>>> a = np.dtype({'names': ['a', 'b'], 'formats': ['i', 'i'], 'offsets': [0, 8]})
>>> b = np.dtype({'names': ['b', 'a'], 'formats': ['i', 'i'], 'offsets': [8, 0]})
>>> a == b
False
```
See how the out of-order offsets in `b` are allowed. In 1.13 the last equality test returned `True`.

aragilar commented:
So in your example, if I have an array with the dtype assigned to `a`, and try to assign it to an array with the dtype assigned to `b`, will numpy copy the buffer across without reordering the underlying buffer (which is what I expect based on the offsets i.e. `'a'` -> `'a'` and `'b'` -> `'b'`), or will numpy shift the contents of the buffer (`'a'` -> `'b'` and `'b'` -> `'a'`).

eric-wieser commented:
It will shift the contents

On Fri, Dec 22, 2017, 00:50 James Tocknell <notifications@github.com> wrote:

> So in your example, if I have an array with the dtype assigned to a, and
> try to assign it to an array with the dtype assigned to b, will numpy
> copy the buffer across without reordering the underlying buffer (which is
> what I expect based on the offsets i.e. 'a' -> 'a' and 'b' -> 'b'), or
> will numpy shift the contents of the buffer ('a' -> 'b' and 'b' -> 'a').
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/6053#issuecomment-353552081>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ9LBwUGOGyzy3PBaoU7eMKSCaHSxJkks5tC21XgaJpZM4FUj1j>
> .
>

aragilar commented:
Is it possible to change the assignment behaviour to prefer using the offsets over the ordering of names before the 1.14 release (h5py doesn't care about matching names, just that the underlying buffer remains the same)? Or is it possible to have subdtypes self-order such that the offsets are aligned?

ahaldane commented:
Well, this change has been planned for more than two years now, as described in previous release notes. If you look in #6053, I have discussion of the many problems and inconsitencies caused by the current behavior, and the motivation for these changes.

It's also not entirely clear to me what you are suggesting. You are currently free to order the fields by offset if you want, so I don't see the point of adding a restriction on the offsets.

ahaldane commented:
Also, is there an issue on the h5py github I can look at were there will be a problem with numpy 1.14? I didn't see anything at first glance.

ahaldane commented:
So I think I should be looking at the fixes in https://github.com/h5py/h5py/pull/793, and at the open issue https://github.com/h5py/h5py/issues/782. It looks like you're still working on the behavior of structured types too, maybe the new 1.14 behavior can be helpful in some way.

Note that the behavior of structured types in 1.13 and before was pretty buggy (see above), which conceivably could have been causing some h5py bugs.

ahaldane commented:
So after looking at it a little, I think the 1.14 changes from this PR actually make the numpy behavior much closer to the internal HD5 internal for compound types. So I think this PR helps you!

It looks like for HDF5 compound datatypes, the order of the fields is as specified by the user, and *not* in the order of the offsets. I checked this by creating a compound type in C using `H5Tinsert`, and if I switch the order I insert the members, the members end up at different member indice, no matter the offsets. That's just like how the order of the fields in a numpy structured type is specified by the user after this PR. I think the fact that "field order matters" is what caused bug https://github.com/h5py/h5py/issues/737 - you weren't maintaining the order of the members.

I also checked the h5py unit test failures. With the caveat that I am very new to the hpy5 code so I might be wrong, I think the two test failures with numpy 1.14 involving "out of order" fields are incorrect: You *should* care about the field order in HDF5, so dtypes with different field orders should *not* be equal. (Where "order" means user-specified order, not offset order)

I also have a comment or two on details of numpy's structured types which I'll add in https://github.com/h5py/h5py/pull/793/.

aragilar commented:
I realised I'd managed to confuse myself about what was happening, if we slightly modify the example to avoid duplicate names:
```
>>> x = np.dtype({'names': ['a', 'b'], 'formats': ['i', 'i'], 'offsets': [0, 8]})
>>> y = np.dtype({'names': ['c', 'd'], 'formats': ['i', 'i'], 'offsets': [8, 0]})
```
I was thinking that `'a'` should map to `'d'`, and `'b'` map to `'c'`, i.e. changing dtype shouldn't affect the underlying buffer. But obviously, changing the dtype will affect the buffer (otherwise why would order matter, and converting from floats to ints wouldn't work). I've set up a tracking bug to fix h5py in at https://github.com/h5py/h5py/issues/969.

Sorry about wasting your time, and thanks for the help (and for releasing RCs, it found the dtype problem and a problem in our CI setup).

eric-wieser commented:
I think you're maybe still not seeing the whole picture here, so:

```python
In [6]: x = np.array((0, 1), dt_x)

In [7]: x
Out[7]:
array((0, 1),
      dtype={'names':['a','b'], 'formats':['<i4','<i4'], 'offsets':[0,8], 'itemsize':12})

In [8]: x.astype(dt_y)  # astype preserves field order (which is what is used in assignment
Out[8]:
array((0, 1),
      dtype={'names':['c','d'], 'formats':['<i4','<i4'], 'offsets':[8,0], 'itemsize':12})

In [9]: x.view(dt_y)  # view preserves buffer order
Out[9]:
array((1, 0),
      dtype={'names':['c','d'], 'formats':['<i4','<i4'], 'offsets':[8,0], 'itemsize':12})
```

aragilar commented:
@eric-wieser Thanks, I'd always assumed indexing with assignment was done via views.

eric-wieser commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1502,6 +1498,35 @@ _get_field_view(PyArrayObject *arr, PyObject *ind, PyArrayObject **view)
+                Py_DECREF(fields);
+                Py_DECREF(names);
+                return 0;
+            }
```
These are new error messages, right? So we're free to change their type in 1.14?

ahaldane replied:
Yes, that's true. Maybe it should have been ValueError here though, because of user's preexisting code like
```
try:
    a[['a', 'c']]
except ValueError:
    return
```
won't work properly. We used to raise ValueError in similar cases, as discussed in #5636.

eric-wieser replied:
I think `ValueError` is actually more appropriate for the second case anyway - `KeyError` implies "key not found" to me, not "that's an invalid value to lookup with"

ihormelnyk commented:
Hi @ahaldane 
Is there any way now to cast an array with a extended dtype to an array with a narrower dtype?
Like it was before 1.14:
```
dt = np.dtype([('a', 'i1'), ('b', 'i4')])
dt_extended = np.dtype([('a', 'i1'), ('zzz', 'i1'), ('b', 'i4')])

data = np.zeros(2, dt_extended)
required_data = np.require(data[:], dtype=dt)
```
Thanks

ahaldane commented:
You need to choose which fields you want to keep, now:
```python
required_data = np.require(data[['a', 'b']], dtype=dt)
```

ihormelnyk commented:
But what if there is nested dtype:
dt = np.dtype([('a', 'i1'), ('b', [('c', 'i4'), ('d', 'f8')])])
dt_extended = np.dtype([('a', 'i1'), ('b', [('c', 'i4'), ('zzz', 'i1'), ('d', 'f8')])])

data = np.zeros(2, dt_extended)
required_data = np.require(data[:], dtype=dt)

Thanks

ahaldane commented:
Hmm, true, there does not seem to be a replacement for that case...

Assignment-by-position still seems a lot better to me than assignment-by-name, but it would be good to find a way to replace all the old use-cases such as that one.

ihormelnyk commented:
Hi @ahaldane ,

Our clients have a lot of python scripts which uses different numpy functions like:
np.require, data.astype(dt2), np.array(data1, dt2)
to extend or to select some subset of data,
dtype is structured(complex) usually.

Sample tests:

```
    def test_squeeze_dtype(self):
        t1 = np.dtype([('a', np.float32), ('b', np.float64)], align=True)
        t2 = np.dtype([('b', np.float64)], align=True)
        data1 = np.zeros(2, t1)
        data2 = np.require(data1, t2)
        data2 = data1.astype(t2)
        data2 = np.array(data1, t2)
        data2 = np.asarray(data1, t2)
        self.assertEqual(data2.dtype.descr, [('b', '<f8')])

    def test_extend_dtype(self):
        t1 = np.dtype([('b', np.float64)], align=True)
        t2 = np.dtype([('a', np.float32), ('b', np.float64)], align=True)
        data1 = np.zeros(2, t1)
        data2 = np.require(data1, t2)
        data2 = data1.astype(t2)
        data2 = np.array(data1, t2)
        data2 = np.asarray(data1, t2)
        self.assertEqual(data2.dtype.descr, [('a', '<f4'), ('', '|V4'), ('b', '<f8')])
```

Error:
```
    return array(a, dtype, copy=False, order=order, subok=True)
ValueError: structures must have the same size
```

We can't upgrade version of numpy for clients higher than 1.13.3 due to "Multiple-field indexing/assignment of structured arrays" changes in 1.14.

Do you plan to implement some functionality in newer version of numpy to cover these old cases?

Thanks,
Ihor Melnyk

ihormelnyk commented:
That would be great

ahaldane commented:
@ihormelnyk  I put up a draft of these functions at #11526.

eric-wieser commented:
@ahaldane: I can think of a non-breaking way to change this functionality - introduce a new `np.structured` data type, which is like `np.void`, but:

* must always have named fields (no fixed-length bytestrings like `void`)
* uses the new semantics

The fact that all numpy dtypes can be structured seems super weird to me - perhaps we could move away from that at the same time as introducing these new semantics

ahaldane commented:
@eric-wieser That would be great. 

My understanding is that the funding @njsmith and others at berkeley got was (in part) to work on the "epic dtype cleanup plan", and I think one of the points there was that structured types should be their own type, not void.

I would strongly support using the new semantics when we create that new type.

eric-wieser commented:
@ihormelnyk's issues look different from any mentioned before. Perhaps the behavior asked for should just be available as `.astype(casting='same_kind')`?

eric-wieser commented:
> We can't upgrade version of numpy for clients higher than 1.13.3 due to "Multiple-field indexing/assignment of structured arrays" changes in 1.14.

AFAIK, these changes were rolled back in 1.14.1 - so you absolutely can have your clients upgrade, as long as they skip 1.14.0

ihormelnyk commented:
@eric-wieser 

```
import sys; print(sys.version)
import numpy; print(numpy.__version__)

import numpy as np

t1 = np.dtype([('a', np.float32), ('b', np.float64)], align=True)
t2 = np.dtype([('b', np.float64)], align=True)
data1 = np.zeros(2, t1)
data2 = np.require(data1, t2)
```
Output:
```
python 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:25:58) [MSC v.1500 64 bit (AMD64)]
numpy 1.14.5

Traceback (most recent call last):
  File "test.py", line 9, in <module>
    data2 = np.require(data1, t2)
  File "Python27\lib\site-packages\numpy\core\numeric.py", line 690, in require
    return asanyarray(a, dtype=dtype)
  File "Python27\lib\site-packages\numpy\core\numeric.py", line 544, in asanyarray
    return array(a, dtype, copy=False, order=order, subok=True)
ValueError: structures must have the same size
```
For numpy 1.13.3 it works well

mattip commented:
For me it works in 1.13.3, but displays
> FutureWarning: Assignment between structured arrays with different field names will change in numpy 1.14.
> Previously fields in the dst would be set to the value of the identically-named field in the src. In numpy 1.14 fields will instead be assigned 'by position': The Nth field of the dst will be set to the Nth field of the src array.

In @ihormelnyk 's case that new behavior would change the dtype `float32` to `float64`.
> See the release notes for details
